src/main.c

```diff
@@ -204,24 +213,70 @@ static const struct xdg_wm_base_listener xdg_wm_base_listener = {
 	.ping = xdg_wm_base_ping,
 };
 
-static void handle_dmabuf_format(void *data,
-		struct zwp_linux_dmabuf_v1 *zwp_linux_dmabuf, uint32_t format)
+static void handle_done(void* data, struct zwp_linux_dmabuf_feedback_v1* zwp_linux_dmabuf_feedback_v1)
 {
 	(void)data;
-	(void)zwp_linux_dmabuf;
+	(void)zwp_linux_dmabuf_feedback_v1;
+
+	if (format_table != MAP_FAILED) {
+		munmap(format_table, format_table_size);
+		format_table = NULL;
+		format_table_size = -1;
+	}
+}
+
+static void handle_format_table(void* data,
+				struct zwp_linux_dmabuf_feedback_v1* zwp_linux_dmabuf_feedback_v1,
+				int32_t fd, uint32_t size)
+{
+	(void)data;
+	(void)zwp_linux_dmabuf_feedback_v1;
+
+	format_table = mmap(NULL, size, PROT_READ, MAP_PRIVATE, fd, 0);
+	if (format_table == MAP_FAILED)
+		return;
+
+	format_table_size = size;
+}
+
+static void handle_tranche_formats(void* data,
+				   struct zwp_linux_dmabuf_feedback_v1* zwp_linux_dmabuf_feedback_v1,
+				   struct wl_array* indices)
+{
+	(void)data;
+	(void)zwp_linux_dmabuf_feedback_v1;
 
 	if (dmabuf_format != DRM_FORMAT_INVALID)
 		return;
 
-	switch (format) {
-	case DRM_FORMAT_XRGB8888:
-	case DRM_FORMAT_XBGR8888:
-		dmabuf_format = format;
+	if (format_table == MAP_FAILED)
+		return;
+
+	uint16_t* index;
+	wl_array_for_each(index, indices) {
+		switch (format_table[*index].format) {
```

Comment:
Although we should be able to assume that the compositor can be trusted, it's probably best to have a sanity check here to make sure that the compositor doesn't send out-of-bounds indices.